torch/csrc/autograd/python_cpp_function.cpp

```diff
@@ -173,22 +173,27 @@ PyObject* functionToPyObject(std::shared_ptr<Function> cdata)
     return obj;
   }
 
-  auto& fn = *cdata;
-  auto it = cpp_function_types.find(std::type_index(typeid(fn)));
-  if (it == cpp_function_types.end()) {
-    return PyErr_Format(PyExc_TypeError,
-        "Don't know how to create Python object for %s", typeid(fn).name());
-  }
+  if (cdata->pyobj) {
+    Py_INCREF(cdata->pyobj);
+  } else {
+    auto& fn = *cdata;
+    auto it = cpp_function_types.find(std::type_index(typeid(fn)));
+    if (it == cpp_function_types.end()) {
+      return PyErr_Format(PyExc_TypeError,
+          "Don't know how to create Python object for %s", typeid(fn).name());
+    }
 
-  PyTypeObject* type = (PyTypeObject*)it->second.get();
-  THPObjectPtr obj(type->tp_alloc(type, 0));
-  if (!obj) return NULL;
-  THPCppFunction* f = (THPCppFunction*)obj.get();
-  new (&f->cdata) std::shared_ptr<Function>(cdata);
-  if (!f->cdata) {
-    return NULL;
+    PyTypeObject* type = (PyTypeObject*)it->second.get();
+    THPObjectPtr obj(type->tp_alloc(type, 0));
+    if (!obj) return NULL;
+    THPCppFunction* f = (THPCppFunction*)obj.get();
+    new (&f->cdata) std::shared_ptr<Function>(cdata);
+
+    cdata->pyobj = obj.release();
+    Py_INCREF(cdata->pyobj);
```

Comment:
Yes, the current implementation has a strong reference because I wasn't sure how the weak ref was done (see [this](https://github.com/pytorch/pytorch/pull/1912#discussion_r128907397) comment).
I did not though that the pyobj are actually our custom object so we can properly remove it on destruction. Thanks for the pointer!